packages/uploads/src/prismaExtension.ts

```diff
@@ -0,0 +1,238 @@
+
+  const resultExtends = {} as ResultExtends
+  for (const modelName in config) {
+    // Guaranteed to have modelConfig, we're looping over config 🙄
+    const modelConfig = config[modelName]
+
+    if (!modelConfig) {
+      continue
+    }
+
+    const uploadFields = (
+      Array.isArray(modelConfig.fields)
+        ? modelConfig.fields
+        : [modelConfig.fields]
+    ) as string[]
+
+    queryExtends[modelName] = {
+      async create({ query, args }) {
+        try {
+          const result = await query(args)
+          return result
+        } catch (e) {
+          // If the create fails, we need to delete the uploaded files
+          await removeUploadedFiles(
+            uploadFields,
+            args.data as Record<string, string>,
+          )
+          throw e
+        }
+      },
+      async update({ query, model, args }) {
+        // Check if any of the uploadFields are present in args.data
+        // We only want to process fields that are being updated
+        const uploadFieldsToUpdate = uploadFields.filter(
+          (field) =>
+            // All of this non-sense is to make typescript happy. I'm not sure how data could be anything but an object
+            typeof args.data === 'object' &&
+            args.data !== null &&
+            field in args.data,
+        )
+
+        // If no upload fields are present, proceed with the original query
+        // avoid overhead of extra lookups
+        if (uploadFieldsToUpdate.length == 0) {
+          return query(args)
+        } else {
+          const originalRecord = await prismaInstance[
+            model as ModelNames
+            // @ts-expect-error TS in strict mode will error due to union type. We cannot narrow it down here.
+          ].findFirstOrThrow({
+            where: args.where,
+            // @TODO: should we select here to reduce the amount of data we're handling
+          })
+
+          // Similar, but not same as create
+          try {
+            const result = await query(args)
+
+            // **After** we've updated the record, we need to delete the old file.
+            await removeUploadedFiles(uploadFieldsToUpdate, originalRecord)
+
+            return result
+          } catch (e) {
+            // If the update fails, we need to delete the newly uploaded files
+            // but not the ones that already exist!
+            await removeUploadedFiles(
+              uploadFieldsToUpdate,
+              args.data as Record<string, string>,
+            )
+            throw e
+          }
+        }
+      },
+
+      async delete({ model, query, args }) {
+        /** Delete args are the same as findFirst, essentially a where clause */
+        const record =
+          // @ts-expect-error TS in strict mode will error due to union type. We cannot narrow it down here.
+          await prismaInstance[model as ModelNames].findFirstOrThrow(args)
+
+        await removeUploadedFiles(uploadFields, record)
```

Comment:
Should this do the same thing as `update` above: run the query first, and only if the query succeeds do you delete the file? I could see a delete failing if there are foreign key constraints...but maybe Prisma handles that check elsewhere and so this code wouldn't even be called if that happened?

Reply:
Ah you're right... I'm sure I added the test case for this too, but maybe i messed up git. I'll fix it1 Thanks!!

Reply:
Updated this now, but the weird thing is.... I tried it without my changes too, and it seems prisma **does** handle the check before it reaches the extension.

Regardless I managed to remove one extra call for `findFirstOrThrow` because delete does return the deleted record, which we need to know what paths to delete!